x/stakeibc/keeper/unbonding_records.go

```diff
@@ -235,6 +245,93 @@ func (k Keeper) GetUnbondingICAMessages(
 	return msgs, unbondings, nil
 }
 
+// In the event that the number of generated undelegate messages exceeds the batch size,
+// reduce the number of messages by dividing any excess amongst proportionally based on
+// the remaining delegation
+// This will no longer be necessary after undelegations to 32+ validators is supported
+func (k Keeper) ConsolidateUnbondingMessages(
+	totalUnbondAmount sdkmath.Int,
+	initialUnbondings []*types.SplitDelegation,
+	unbondCapacities []ValidatorUnbondCapacity,
+	batchSize int,
+) (finalUnbondings []*types.SplitDelegation, err error) {
+	// Grab the first {batch_size} number of messages from the list
+	// This will consist of the validators with the most capacity
+	unbondingsCapped := initialUnbondings[:batchSize]
+
+	// Calculate the amount that was initially meant to be unbonded from that batch,
+	// and determine the remainder that needs to be redistributed
+	unbondAmountFromBatch := sdkmath.ZeroInt()
+	initialUnbondingsMap := map[string]sdkmath.Int{}
+	for _, unbonding := range unbondingsCapped {
+		unbondAmountFromBatch = unbondAmountFromBatch.Add(unbonding.Amount)
+		initialUnbondingsMap[unbonding.Validator] = unbonding.Amount
+	}
+	totalExcessAmount := totalUnbondAmount.Sub(unbondAmountFromBatch)
+
+	// Store the delegation of each validator that was expected *after* the originally
+	// planned unbonding went through
+	// e.g. If the validator had 10 before unbonding, and in the first pass, 3 was
+	//      supposed to be unbonded, their delegation after the first pass is 7
+	totalDelegationsAfterFirstPass := sdk.ZeroDec()
+	delegationsAfterFirstPass := map[string]sdk.Dec{}
+	for _, capacity := range unbondCapacities {
+		// Only add validators that were in the initial unbonding plan
+		// The delegation after the first pass is calculated by taking the "current delegation"
+		// (aka delegation before unbonding) and subtracting the unbond amount
+		if initialUnbondAmount, ok := initialUnbondingsMap[capacity.ValidatorAddress]; ok {
+			delegationAfterFirstPass := sdk.NewDecFromInt(capacity.CurrentDelegation.Sub(initialUnbondAmount))
+
+			delegationsAfterFirstPass[capacity.ValidatorAddress] = delegationAfterFirstPass
+			totalDelegationsAfterFirstPass = totalDelegationsAfterFirstPass.Add(delegationAfterFirstPass)
+		}
+	}
+
+	// This should never happen, but it protects against a division by zero error below
+	if totalDelegationsAfterFirstPass.IsZero() {
```

Comment:
ah actually, I was misunderstand this when I wrote it (I previously thought it was impossible, but it's actually possible just unlikely)

This would occur if the 32 validators with the most capacity also happened to all be validators with 0 weight. Cause in that case, we would be unbonding everything from them

Reply:
oh, this is an interesting case.  Even if not all 32 would be zero weight validators, there could be enough in the top 32 that the non-zero weight validators get all of their delegation unbonded... probably unlikely in practice without monster LSMs happening frequently.

But hold on, we wouldn't necessarily be unbonding everything from even the zero weight ones, it is just that for them capacity == currentDelegation because balancedDelegation = 0.  It would depend on the total unbonding amount across all the messages if it is greater than the delegations across all the zero weight ones but I get what you are saying.